### ISSUE TITLE
Compound open statement

### ISSUE THREAD

**Issue by duncanmak** [fsharp/fslang-suggestions#657]
# Compound `open` statement

I propose we introduce a more succinct syntax for writing open statements, one that allows the opening of multiple namespaces with the same root namespace in a single line.

The existing way of approaching this problem in F# is ...

As an example, from fsharp's own `ast.fs`, we see:

```F#
open System.Collections.Generic
open Internal.Utilities.Text.Lexing
open Internal.Utilities.Text.Parsing
open Microsoft.FSharp.Compiler.AbstractIL
open Microsoft.FSharp.Compiler.AbstractIL.IL
open Microsoft.FSharp.Compiler.AbstractIL.Internal
open Microsoft.FSharp.Compiler.AbstractIL.Internal.Library
open Microsoft.FSharp.Compiler
open Microsoft.FSharp.Compiler.UnicodeLexing
open Microsoft.FSharp.Compiler.ErrorLogger
open Microsoft.FSharp.Compiler.PrettyNaming
open Microsoft.FSharp.Compiler.Range
```

The repetition could be eliminated if we allow

```F#
open System.Collections.Generic
open Internal.Utilities.Text.(Lexing, Parsing)
open Microsoft.FSharp.Compiler.AbstractIL.(_, IL, Internal, Internal.Library)
open Microsoft.FSharp.Compiler.(_, UnicodeLexing, ErrorLogger, PrettyNaming, Range)
```

#### The advantages of making this adjustment to F# are:

There's less to type, this come particularly handy in the scripting case, or when using `fsharpi`.

#### The disadvantages of making this adjustment to F# are:

The simple explanation of this syntax is that

`open A.(B, C)` expands out to `open A.B` and `open A.C`. This is simple to learn and explain.

 However, there's an issue regarding `open A`.

To allow `open A` to also be part of the expansion, a _null string_ identifier is needed. `UNDERSCORE` and `DOT` are two candidates for this identifier.

This identifier is not self-evident, if the person writing the code is not aware of this identifier, they might be surprised that `open A.(B, C)` does not `open A` as well.

Instead, they have to write either
```
open A
open A.(B, C)
```
or learn the identifier and write `open A.(_, B, C)`

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): S

There are two open design issues:

1. What's the _null string_ identifier? `DOT` and `UNDERSCORE` have been suggested. 

An alternative: Should the base namespace always be opened? After all, this is just a shorthand syntax, if someone wants `B` and `C` open, but not `A`, they could always write that out explicitly in the longer form.

2. Do we allow nesting of these parenthesized paths? If we do, the example above can be further shorten as `AbstractIL` could also be condensed into the parenthesized list. i.e. 
```
open Microsoft.FSharp.Compiler.
        (_,
         AbstractIL.(_, IL, Internal, Internal.Library)
         UnicodeLexing, ErrorLogger, PrettyNaming, Range)
```

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

**Comment by auduchinok** [comment]
To be honest, I don't see much point in this syntax, moreover I find it **much** more difficult to reason about at a glance, despite the code being more compact.

**Comment by TIHan** [comment]
@duncanmak Thank you for writing this up as we have a place to talk about it.

I'm not quite in favor of the suggestion as it seems there are more disadvantages than there are advantages. It adds another layer of complexity in the language and tooling just to save some typing; which an editor can help you out anyways. That is just my personal opinion.

However, I don't want to discourage you from trying to implement this feature as an experiment/exercise. I think it's a great way to get familiar with the compiler.

**Comment by piaste** [comment]
I've never liked massive `open` blurbs and I'd welcome some effort towards making them more readable at a glance.

That said, I agree that having to learn a special syntax for this would be hard to justify.

**I suggest we make the syntax as similar as possible to existing F# syntax**, to minimize confusion. For example, let's say indentation means 'prepend this namespace to the indented namespaces':

```
open System.Collections.Generic
open Internal.Utilities.Text
  open Lexing
  open Parsing
open Microsoft.FSharp.Compiler.AbstractIL
  open IL
  open Internal
    open Library
open Microsoft.FSharp.Compiler
  open UnicodeLexing
  open ErrorLogger
  open PrettyNaming
  open Range
```

And for consistency with the rest of the language, it would make sense to have an equivalent 'verbose' syntax. This would allow one to save vertical space at the expense of cleanliness, which I think is what @duncanmak was shooting for. Let's say the verbose keywords could be `open .. with .. end`:

```
open System.Collections.Generic end
open Internal.Utilities.Text with open Lexing; open Parsing; end
open Microsoft.FSharp.Compiler.AbstractIL with open IL; open Internal with open Library; end; end
open Microsoft.FSharp.Compiler with 
  open UnicodeLexing
  open ErrorLogger
  open PrettyNaming
  open Range
end
```

Note that this syntax doesn't cover the case where you want to open multiple several namespaces, but explicitly do not want to open their common parent. I think it's an uncommon enough case to be worth spelling out explicitly when it happens.

**Comment by abelbraaksma** [comment]
@piaste, you wrote exactly what I had in mind to write in this thread (except that I didn't get around to doing it yet ;)).

I agree that any syntax should look like existing syntax that people are familiar with. New syntax that includes the introduction or parens is only going to add to the clutter. Furthermore, horizontal reading is harder for most users, vertical reading is faster and cleaner.

Your nested proposal makes sense. It is clear at a glance what belongs to what. It should be mentioned that if the vertical space is precious to people, that they can already use the semicolon syntax:

```f#
open System; open System.Xml.Linq;open Microsoft.FSharp.Reflection  // legal
```

It should also be mentioned that your (@piaste) first nested syntax is presently legal. That means that introducing it like you propose is going to be backwards incompatible:

```f#
open System.Collections.Generic
open Internal.Utilities.Text
  open Lexing   // currently legal, it matches any top-level "Lexing" namespace, this syntax changes that
  open Parsing
```

In other words: nesting is already possible and would aid in readability, but you (currently) have to repeat the whole namespace:

```f#
open System.Collections.Generic
open Internal.Utilities.Text
  open Internal.Utilities.Text.Lexing
  open Internal.Utilities.Text.Parsing
```

I would like to suggest a slightly different approach which doesn't break existing code (I think). Basically, I would like to combine your with-syntax and your indent-syntax into one. Consider, for instance:

```f#
open System.Collections.Generic
open Internal.Utilities.Text as txt with
  | txt.Lexing
  | txt.Parsing
open Microsoft.FSharp.Compiler.AbstractIL abs with
  | abs.IL
  | abs.Internal as int with
      | int.Library
```

This has the advantage that it will help any intellisense and syntax highlighting and may be easier to parse. Also, the `as ... with` resembles syntax we already know and parsing for `x... with... | ....|` already exists, which may make this easier to tokenize.

You may argue that this adds repetition (the `as X` syntax), but it may also aid the eye in understanding the meaning of scope and more cleanly shows the intention of the programmer. 

Just like with regular matching syntax, people who like to write them horizontally could do so:

```f#
open System.Collections.Generic
open Internal.Utilities.Text as txt with | txt.Lexing | txt.Parsing
open Microsoft.FSharp.Compiler.AbstractIL abs with | abs.IL | abs.Internal as int with | int.Library
```

If you don't like to write `as X`, we could consider the same as with class syntax:

```f#
open System.Collections.Generic
open Internal.Utilities.Text with
  | __.Lexing
  | __.Parsing
open Microsoft.FSharp.Compiler.AbstractIL with
  | __.IL
  | __.Internal with
      | __.Library
```

Personally, the added `|` in comparison to your proposal adds clarity (but also a tad of extra clutter), but more importantly, it doesn't clash with existing, legal syntax.

Whether or not the scoping rules with implicit/explicit `as` should be used, I don't know. I think it is cleaner then, say, this:

```f#
open System.Collections.Generic
open Internal.Utilities.Text with
  Lexing
  Parsing
open Microsoft.FSharp.Compiler.AbstractIL with
  IL
  Internal with
      Library
```

Or this:

```f#
open System.Collections.Generic
open Internal.Utilities.Text with
  .Lexing
  .Parsing
open Microsoft.FSharp.Compiler.AbstractIL with
  .IL
  .Internal with
      .Library
```

Or this:

```f#
open System.Collections.Generic
open Internal.Utilities.Text with
  open Lexing
  open Parsing
open Microsoft.FSharp.Compiler.AbstractIL with
  open IL
  open Internal with
      open Library
```

**Comment by lambdageek** [comment]
I like @abelbraaksma last proposal except that I think all the `open`s except the first one shouldn't be there:
```fsharp
open System.Collections.Generic
open Internal.Utilities.Text with
  Lexing
  Parsing
open Microsoft.FSharp.Compiler.AbstractIL with
  IL
  Internal with
    Library
```

single line using `;` and `end` is also possible to mix with multiline:

```fsharp
open System.Collections.Generic
open Internal.Utilities.Text with Lexing; Parsing; end
open Microsoft.FSharp.Compiler.AbstractIL with
  IL
  Internal with Library; end
```

**Comment by cartermp** [comment]
I'm in favor of this change and think it would be a good fit overall for the language assuming that tooling-level work is also done. The other syntax alternatives mentioned here don't seem to offer much in terms of typing less code, which is the primary reason why this suggestion exists.

To that end, I would support the following:

```fsharp
open System
open System.IO
open System.Linq
open System.Threading.Tasks
```

Is equivalent to:

```fsharp
open System
open System.(IO, Linq, Threading.Tasks)
```

Discard values or implicit opens would be out of scope in my mind. That is, I would **not** support the following at this time:

```fsharp
// 'System' is implicitly openened
// Noooo
open System.(IO, Linq, Threading.Tasks)

// '_' makes the most sense since it is used in wildcard situations
// But nooooooo
open System.(_, IO, Linq, Threading.Tasks)
```

I'd be more open to requiring a `_` than making it implicit. But even then I just don't like the wildcard character, no matter which form it takes.

From a tooling perspective, I would expect the following:

Today, "normal" directives are the same:
```fsharp
open System. //<<- Normal completion here
```
With this change:

```fsharp
open System.( //<<- Same completion as in previous example
```
```fsharp
open System.(Dia //<<- Filtered completion list with 'Diagnostics' preselected
```
```fsharp
open System.(Threading. // Same completion as if I typed 'System.Threading.'
```

Additionally, I am open to nested compound opens. That is, the following would work: 

```fsharp
open System
open System.Collections.Generic
open System.Collections.Concurrent
open System.Collections.Immutable
```

Becomes:

```fsharp
open System
open System.(Collections.(Generic, Concurrent, Immutable))
```

**Comment by cartermp** [comment]
It's also worth pointing out that there is a similar proposal for Rust: https://github.com/rust-lang/rfcs/blob/master/text/2128-use-nested-groups.md

And Scala supports a form of this: https://docs.scala-lang.org/tour/packages-and-imports.html

**Comment by 0x53A** [comment]
Rust already supports a different feature that looks at first glance similar. It seems like the linked proposal is an extension of that?

But this is often used to __limit__ the number of imports, not just save a few keystrokes:

Normally you open a namespace this way:

```rust
use namespace1::namespace2::*;
```

If there are a ton of types, but you only need one or two, you can specifically import only them:

```rust
use namespace1::namespace2::{Type1, Type2, Type3};
```

So the equivalent F# would be

```F#
open System.Generics.{List, HashSet}
```

**Comment by cartermp** [comment]
@0x53A I believe the proposal is an extension of this, yes. Like you can do on the JVM, they support limiting the scope of what's imported. I think that's a separate issue which is worth a suggestion, IMO.

**Comment by duncanmak** [comment]
I just updated https://github.com/Microsoft/visualfsharp/pull/4673 to align with @cartermp's suggestions.

- The underscore thing was reverted,
- Nested compound opens are supported, so this works:
```fsharp
open System.(Collections.(Generic, Concurrent, Specialized))
```

**Comment by tytusse** [comment]
How about splitting statements to multiple lines with indentation? Say, something like this:
```
open 
   System
      *
      Collections
         Generic
         Concurrent
         Specialized
```
would be equivalent to:
```FSharp
open System
open System.Collections.Generic
open System.Collections.Concurrent
open System.Collections.Specialized
```

General idea would be to have more "visual" style than the one with brackets. Also I am not sure about `*` symbol but I have no idea for alternative. `*` seems out of place here - usually it represents some kind of wildcard resolving to "everything in" - in this case - everything in "System" namespace.

Also I am aware that original proposal contains "... in a single line..." statement ;-)

**Comment by realvictorprm** [comment]
@tytusse That's an interesting idea. It feels to me fresh and natural to use indentation  for such.

**Comment by jwosty** [comment]
@tytusse Interesting idea, but it seems to have backfired -- that snippet is now longer (takes 7 LOC) 😕

**Comment by duncanmak** [comment]
@jwosty that's exactly my reaction to proposals like @tytusse's too.

One use that I'm definitely targeting is the `fsi` interactive top-level. It'd be nice to be able to get your namespaces imported without too much ceremony.

Here is a suggestion from @lambdageek that I liked: use a different separate apart from `DOT` to denote whether or not to also open the parent.

To use @tytusse's example:

> ```fsharp
> open System.Collections.(Generic, Concurrent, Specialized)
> ```
> in this case, it'll not `open System`.

Using a different separator (in this case, `COMMA`):

> ```fsharp
> open System:Collections.(Generic, Concurrent, Specialized)
> ```
>
> in this case, do `open System`.

The new separator could be anything - `:`, `$`, `%`, etc...

The case against this is of course quite obvious - it's new syntax, it's not a convention in any other .NET language (though maybe once this gets merged, the natural next step is to get a similar change into Roslyn for C#....).

What do you think, @dsyme @cartermp?

P.S. The compiler side of this PR is done, but I need to find time to update the tooling side. If anyone's interested in helping out here, it'd be much appreciated.

**Comment by jwosty** [comment]
I do like the `open System.Collections.(Generic, Concurrent, Specialized)` syntax (or as @cartermp suggested, `open System.(Collections.(Generic, Concurrent, Specialized))`). Seems pretty clean to me.

I personally wouldn't get super worried about being able to represent arbitrary `open` hierarchies with this, though; if it gets to that point, the user is probably overdoing it. For example, if I'm doing all this:

```fsharp
open System
open System.Collections.Generic
open System.Collections.Concurrent
open System.Collections.Immutable
open System.Runtime.InteropServices
open System.Runtime.CompilerServices
open System.Runtime.Versioning
```

it would be strong ill-advised to write this:

```fsharp
open System.(Collections.(Generic, Concurrent, Immutable)).(Runtime.(InteropServices, CompilerServices, Versioning))
```

when this suffices to make it shorter but keeping it readable and easy to digest

```fsharp
open System
open System.(Collections.(Generic, Concurrent, Immutable))
open System.(Runtime.(InteropServices, CompilerServices, Versioning))
```

In other words, I don't care too much for the compiler allowing us to cram huge hierarchies into an obfuscated mess suitable. Unless I'm code-golfing... ;)

**Comment by matthid** [comment]
Honestly, I'm not sure this feature is worth it - considering the IDE should just generate the open statements for you. We should work towards making that work more reliable across all the IDEs we have.
In a perfect world I don't even want to write the open statements manually and I don't expect my IDE to group them magically.

Also ordering sometimes matters especially when you don't open fully qualified namespaces (and there is a name class) or import operators, so you might not even be able to use this particular feature on some occasions which might make it even less useful then we think.

**Comment by dsyme** [comment]
I've not commented much on this feature, mainly because I wanted to let it run and see where it went.

**Comment by duncanmak** [comment]
I'll defer to the community and @dsyme to decide whether or not this feature should make it in.

I'll just say that this feature came to me when I was using `fsi` a lot - I agree with @matthid that inside an IDE, this is not as useful, but in a scripting scenario, it's definitely handy. 

This is also not any new innovation, a similar feature exists in Scala, Rust, Clojure, etc, etc.

I'm bringing this up now again because of @dsyme's work on https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1068-open-static-classes.md. To me, these two proposals are related in that they modify the `open` statement to makes it less verbose to bring in identifiers from namespaces/modules.

**Comment by dsyme** [comment]
I will close this for the reasons outlined here: https://github.com/fsharp/fslang-suggestions/issues/657#issuecomment-472552153
